File: backend/services/groq_api.py
import os
from groq import Groq
import json
from dotenv import load_dotenv

load_dotenv()

# Lazy initialization of Groq client
_groq_client = None

def get_groq_client():
    """
    Get or create the Groq client instance.
    Uses lazy initialization to ensure .env is loaded first.
    """
    global _groq_client
    
    if _groq_client is None:
        groq_api_key = os.getenv('GROQ_API_KEY')
        
        if groq_api_key:
            _groq_client = Groq(api_key=groq_api_key)
        else:
            print("Warning: GROQ_API_KEY not found in environment variables")
            return None
    
    return _groq_client

def generate_restaurant_tags(restaurant):
    """
    Generate tags for a restaurant using Groq AI.
    
    Args:
        restaurant (dict): Restaurant object from Google Places
    
    Returns:
        list: List of generated tags
    """
    client = get_groq_client()
    if not client:
        return ['restaurant']  # Default fallback tag
    
    try:
        restaurant_name = restaurant.get('name', 'Unknown')
        restaurant_types = restaurant.get('types', [])
        
        # Create a prompt for the AI
        prompt = f"""Given a restaurant named "{restaurant_name}" with the following types: {', '.join(restaurant_types)}, 
generate 3-5 relevant tags that describe this restaurant. The tags should be short, descriptive words or phrases.

Return ONLY a JSON array of strings, nothing else. Example: ["italian", "fine dining", "romantic"]"""
        
        # Use llama-3.3-70b-versatile (free tier model)
        chat_completion = client.chat.completions.create(
            messages=[
                {
                    "role": "user",
                    "content": prompt,
                }
            ],
            model="llama-3.3-70b-versatile",  # Free tier model
            temperature=0.7,
            max_tokens=100,
        )
        
        response_text = chat_completion.choices[0].message.content.strip()
        
        # Try to parse the response as JSON
        try:
            tags = json.loads(response_text)
            if isinstance(tags, list) and all(isinstance(tag, str) for tag in tags):
                return tags
        except json.JSONDecodeError:
            pass
        
        # Fallback: extract tags from response
        return extract_tags_from_text(response_text)
    
    except Exception as e:
        print(f"Error generating tags with Groq: {e}")
        return ['restaurant']  # Default fallback

def extract_tags_from_text(text):
    """
    Extract tags from text response if JSON parsing fails.
    
    Args:
        text (str): Response text
    
    Returns:
        list: List of extracted tags
    """
    # Simple extraction: look for words in quotes or comma-separated values
    tags = []
    
    # Try to find quoted strings
    import re
    quoted = re.findall(r'"([^"]*)"', text)
    if quoted:
        tags = quoted[:5]  # Limit to 5 tags
    
    if not tags:
        # Fallback to splitting by commas
        parts = text.replace('[', '').replace(']', '').split(',')
        tags = [part.strip().strip('"').strip("'") for part in parts if part.strip()][:5]
    
    return tags if tags else ['restaurant']

File: api/_lib/services/groqApi.ts
import Groq from 'groq-sdk';
import type { GooglePlacesRestaurant } from './googlePlaces';

let groqClient: Groq | null = null;

function getGroqClient(): Groq | null {
  if (!groqClient) {
    const apiKey = process.env.GROQ_API_KEY;
    if (!apiKey) {
      console.warn('GROQ_API_KEY not found in environment variables');
      return null;
    }
    groqClient = new Groq({ apiKey });
  }
  return groqClient;
}

function normalizeTags(tags: string[]): string[] {
  return tags.map(tag =>
    tag.toLowerCase().trim().replace(/_/g, ' ').replace(/\s+/g, ' ')
  );
}

function extractTagsFromText(text: string): string[] {
  const quoted = text.match(/"([^"]*)"/g)?.map(s => s.replace(/"/g, '')) ?? [];
  if (quoted.length) return normalizeTags(quoted.slice(0, 5));

  const parts = text.replace(/[\[\]]/g, '').split(',').map(s => s.trim().replace(/^['"]|['"]$/g, '')).filter(Boolean);
  return parts.length ? normalizeTags(parts.slice(0, 5)) : ['restaurant'];
}

const CANONICAL_TAGS = [
  // Cuisine
  'american', 'italian', 'mexican', 'chinese', 'japanese', 'korean', 'thai', 'indian',
  'mediterranean', 'greek', 'french', 'spanish', 'vietnamese', 'asian', 'middle eastern',
  'latin', 'caribbean', 'african', 'fusion',
  // Food type
  'pizza', 'sushi', 'burgers', 'tacos', 'sandwiches', 'salads', 'seafood', 'steakhouse',
  'bbq', 'wings', 'noodles', 'ramen', 'pho', 'dumplings', 'burritos', 'pasta', 'breakfast',
  'brunch', 'desserts', 'bakery', 'coffee', 'smoothies', 'ice cream',
  // Dining style
  'fast food', 'casual dining', 'fine dining', 'cafe', 'food truck', 'buffet', 'bar and grill',
  'sports bar', 'hotel restaurant', 'diner', 'bistro', 'takeout', 'delivery',
  // Vibe / attributes
  'family friendly', 'date night', 'late night', 'healthy', 'vegetarian', 'vegan',
  'halal', 'kosher', 'outdoor seating', 'upscale', 'quick bites',
];

export async function generateRestaurantTags(restaurant: GooglePlacesRestaurant): Promise<string[]> {
  const client = getGroqClient();
  if (!client) return ['restaurant'];

  try {
    const prompt = `Given a restaurant named "${restaurant.name}" with types: ${restaurant.types.join(', ')}, choose 3-5 tags from this list that best describe it:

${CANONICAL_TAGS.join(', ')}

Rules:
- Only use tags from the list above
- Return ONLY a JSON array of strings, no explanation
- Example: ["italian", "pizza", "casual dining"]`;

    const completion = await client.chat.completions.create({
      model: 'llama-3.3-70b-versatile',
      temperature: 0.7,
      max_tokens: 100,
      messages: [{ role: 'user', content: prompt }],
    });

    const text = completion.choices[0].message.content?.trim() ?? '';

    try {
      const tags = JSON.parse(text);
      if (Array.isArray(tags) && tags.every(t => typeof t === 'string')) return normalizeTags(tags);
    } catch {
      // fall through to text extraction
    }

    return extractTagsFromText(text);
  } catch (e) {
    console.error('Error generating tags with Groq:', e);
    return ['restaurant'];
  }
}
